// sucrase/dist/cli.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }/* eslint-disable no-console */
var _commander = require('commander'); var _commander2 = _interopRequireDefault(_commander);
var _glob = require('glob');
var _fs = require('mz/fs');
var _path = require('path');

var _index = require('./index');











 function run() {
  _commander2.default
    .description(`Sucrase: super-fast Babel alternative.`)
    .usage("[options] <srcDir>")
    .option(
      "-d, --out-dir <out>",
      "Compile an input directory of modules into an output directory.",
    )
    .option(
      "-p, --project <dir>",
      "Compile a TypeScript project, will read from tsconfig.json in <dir>",
    )
    .option("--out-extension <extension>", "File extension to use for all output files.", "js")
    .option("--exclude-dirs <paths>", "Names of directories that should not be traversed.")
    .option("-q, --quiet", "Don't print the names of converted files.")
    .option("-t, --transforms <transforms>", "Comma-separated list of transforms to run.")
    .option("--disable-es-transforms", "Opt out of all ES syntax transforms.")
    .option("--jsx-runtime <string>", "Transformation mode for the JSX transform.")
    .option("--production", "Disable debugging information from JSX in output.")
    .option(
      "--jsx-import-source <string>",
      "Automatic JSX transform import path prefix, defaults to `React.Fragment`.",
    )
    .option(
      "--jsx-pragma <string>",
      "Classic JSX transform element creation function, defaults to `React.createElement`.",
    )
    .option(
      "--jsx-fragment-pragma <string>",
      "Classic JSX transform fragment component, defaults to `React.Fragment`.",
    )
    .option("--keep-unused-imports", "Disable automatic removal of type-only imports/exports.")
    .option("--preserve-dynamic-import", "Don't transpile dynamic import() to require.")
    .option(
      "--inject-create-require-for-import-require",
      "Use `createRequire` when transpiling TS `import = require` to ESM.",
    )
    .option(
      "--enable-legacy-typescript-module-interop",
      "Use default TypeScript ESM/CJS interop strategy.",
    )
    .option("--enable-legacy-babel5-module-interop", "Use Babel 5 ESM/CJS interop strategy.")
    .parse(process.argv);

  if (_commander2.default.project) {
    if (
      _commander2.default.outDir ||
      _commander2.default.transforms ||
      _commander2.default.args[0] ||
      _commander2.default.enableLegacyTypescriptModuleInterop
    ) {
      console.error(
        "If TypeScript project is specified, out directory, transforms, source " +
          "directory, and --enable-legacy-typescript-module-interop may not be specified.",
      );
      process.exit(1);
    }
  } else {
    if (!_commander2.default.outDir) {
      console.error("Out directory is required");
      process.exit(1);
    }

    if (!_commander2.default.transforms) {
      console.error("Transforms option is required.");
      process.exit(1);
    }

    if (!_commander2.default.args[0]) {
      console.error("Source directory is required.");
      process.exit(1);
    }
  }

  const options = {
    outDirPath: _commander2.default.outDir,
    srcDirPath: _commander2.default.args[0],
    project: _commander2.default.project,
    outExtension: _commander2.default.outExtension,
    excludeDirs: _commander2.default.excludeDirs ? _commander2.default.excludeDirs.split(",") : [],
    quiet: _commander2.default.quiet,
    sucraseOptions: {
      transforms: _commander2.default.transforms ? _commander2.default.transforms.split(",") : [],
      disableESTransforms: _commander2.default.disableEsTransforms,
      jsxRuntime: _commander2.default.jsxRuntime,
      production: _commander2.default.production,
      jsxImportSource: _commander2.default.jsxImportSource,
      jsxPragma: _commander2.default.jsxPragma || "React.createElement",
      jsxFragmentPragma: _commander2.default.jsxFragmentPragma || "React.Fragment",
      keepUnusedImports: _commander2.default.keepUnusedImports,
      preserveDynamicImport: _commander2.default.preserveDynamicImport,
      injectCreateRequireForImportRequire: _commander2.default.injectCreateRequireForImportRequire,
      enableLegacyTypeScriptModuleInterop: _commander2.default.enableLegacyTypescriptModuleInterop,
      enableLegacyBabel5ModuleInterop: _commander2.default.enableLegacyBabel5ModuleInterop,
    },
  };

  buildDirectory(options).catch((e) => {
    process.exitCode = 1;
    console.error(e);
  });
} exports.default = run;






async function findFiles(options) {
  const outDirPath = options.outDirPath;
  const srcDirPath = options.srcDirPath;

  const extensions = options.sucraseOptions.transforms.includes("typescript")
    ? [".ts", ".tsx"]
    : [".js", ".jsx"];

  if (!(await _fs.exists.call(void 0, outDirPath))) {
    await _fs.mkdir.call(void 0, outDirPath);
  }

  const outArr = [];
  for (const child of await _fs.readdir.call(void 0, srcDirPath)) {
    if (["node_modules", ".git"].includes(child) || options.excludeDirs.includes(child)) {
      continue;
    }
    const srcChildPath = _path.join.call(void 0, srcDirPath, child);
    const outChildPath = _path.join.call(void 0, outDirPath, child);
    if ((await _fs.stat.call(void 0, srcChildPath)).isDirectory()) {
      const innerOptions = {...options};
      innerOptions.srcDirPath = srcChildPath;
      innerOptions.outDirPath = outChildPath;
      const innerFiles = await findFiles(innerOptions);
      outArr.push(...innerFiles);
    } else if (extensions.some((ext) => srcChildPath.endsWith(ext))) {
      const outPath = outChildPath.replace(/\.\w+$/, `.${options.outExtension}`);
      outArr.push({
        srcPath: srcChildPath,
        outPath,
      });
    }
  }

  return outArr;
}

async function runGlob(options) {
  const tsConfigPath = _path.join.call(void 0, options.project, "tsconfig.json");

  let str;
  try {
    str = await _fs.readFile.call(void 0, tsConfigPath, "utf8");
  } catch (err) {
    console.error("Could not find project tsconfig.json");
    console.error(`  --project=${options.project}`);
    console.error(err);
    process.exit(1);
  }
  const json = JSON.parse(str);

  const foundFiles = [];

  const files = json.files;
  const include = json.include;

  const absProject = _path.join.call(void 0, process.cwd(), options.project);
  const outDirs = [];

  if (!(await _fs.exists.call(void 0, options.outDirPath))) {
    await _fs.mkdir.call(void 0, options.outDirPath);
  }

  if (files) {
    for (const file of files) {
      if (file.endsWith(".d.ts")) {
        continue;
      }
      if (!file.endsWith(".ts") && !file.endsWith(".js")) {
        continue;
      }

      const srcFile = _path.join.call(void 0, absProject, file);
      const outFile = _path.join.call(void 0, options.outDirPath, file);
      const outPath = outFile.replace(/\.\w+$/, `.${options.outExtension}`);

      const outDir = _path.dirname.call(void 0, outPath);
      if (!outDirs.includes(outDir)) {
        outDirs.push(outDir);
      }

      foundFiles.push({
        srcPath: srcFile,
        outPath,
      });
    }
  }
  if (include) {
    for (const pattern of include) {
      const globFiles = await _glob.glob.call(void 0, _path.join.call(void 0, absProject, pattern));
      for (const file of globFiles) {
        if (!file.endsWith(".ts") && !file.endsWith(".js")) {
          continue;
        }
        if (file.endsWith(".d.ts")) {
          continue;
        }

        const relativeFile = _path.relative.call(void 0, absProject, file);
        const outFile = _path.join.call(void 0, options.outDirPath, relativeFile);
        const outPath = outFile.replace(/\.\w+$/, `.${options.outExtension}`);

        const outDir = _path.dirname.call(void 0, outPath);
        if (!outDirs.includes(outDir)) {
          outDirs.push(outDir);
        }

        foundFiles.push({
          srcPath: file,
          outPath,
        });
      }
    }
  }

  for (const outDirPath of outDirs) {
    if (!(await _fs.exists.call(void 0, outDirPath))âƒÚ‰#XD(`` ⁄L0.’<EôØ	¯\ñF‚ÒΩí∂j $ ø«3”÷C÷DÃÒ!eƒ rµTú∆k)
ÄıècÇ™Ä9C7ï^†Î€i–Vpµ»»ÑR2¨ïTÑ rÿëIä1
$EÂ
	`’ùÔêêP£™Ã.2ÈíQí∏™#ê4Ø»;¿aΩX :‡J°$C"ãπ):ﬁË"(°?ÏyD'0Å	ZÄÇ¿0ÈcÖ :äƒ!)æ˙x¯Hê",
Ï6`≤¯ c6lä‰mNí5"Ü'63 —œÚ.√	N‚WP‹PÑ	¯  ¯(iP Eáãu‚¡y»®ë H»hLájy≥XUàïàƒDôv9Ç"?Ä_ ˇ≤
&∑" ï6∞JƒVCC†ñ∆Ä«≤Ó%lå†0éΩ—E!'A¿‚¿$pa¥KÇ¥è›í-÷*N°Sê=	
¿†I ”
Y∞Pb
%∏4b<Í… `¥‡≥∑«©yWŸr≥\<Ñ0TG‹@≤®ëÅ'7@1XOåFÒ¬Lëo1)30Å,™Ë€Ì{Í∑{FlQ@›nòôB?$ê∏8	"S0M6zIõ0PqJ’eá1u k D¢sÜ:Œú+„˙&]…}g?ÄX£‰+ç˘HS8[Ü>ê∏"çGU
KF5,0@xÇ#«mJ@%≤hIXΩÄ∞#Û‚\o††m	ÅF√0$p!â*0Ω•¯ËTÄb5
0  Øe&B ™â~HZìz≤ÙÇr3
Ï$4à¬±·πd[y≠xìo9∆äZW«úç◊p€
p'ıuˇï–ÜË„UtníìWµÒ¿˜O!R•Üì’ÿ!6HÇ¿’t°»‰_cFí
{ì|8µÆ$!Ä
âÅÆ
1a"cï,î©( …œrCÅ‰G:`I˝È•!yÿHRöc
«P)÷˝ç¥Sπ¡Jû-P~âÇ(—](±PÒÙ €Ö(ö–k
õöE6«»VÈ)ã≠"ÆR#ÂªÎI¢ìS:Å)DÿDJú 	;«kX‘gÌ‰Fc‹RÜcõJ>Ÿ
W„⁄ÊA
Ü`√‰akúSLAs≤5D–Èg©R¢0Ç/ùÜ	Ó'ítZ4ÙC`¬`J” "çj•»aÆ pzÒ,`ª-ë
úÇD™Ö5 Ç Pòø&∫1Û†
»áÈà5\W≤‡¸@@aÄÒœÚH≠€#Çà∞@ÅÍ'Ä!Äg¿#£ÃK%»(f≠âSπvD@ííä∏•∞Å†¿á{ÊÙíM∞ (©WH0¶á•CéAbå≈ß>‘Ö'-˝∞Gë˛pâ…”œ24Êw±é`h+í§ùòåB	î‡Ç ¬áË©ùü+gHœ™ ¯ËäÜ_¿2WZò4˚ÓÜŒ©#° ÿó==∞+W‰bƒ”+\ÑÑàú¡MòÖ X†ÿ)j∂|Tî˙Ò0Æ¢¬Íàt{>w®MU÷Ä¡ ¡±Âæ∆rR· áõhÈ(F‰ „!%ƒ"XT3 \tÛ»AQ'¡Y™˝(j;1@63ËÄXöÔl»<T+`í#Z)Dréù√.äé 0L˝z' VÄÙ¬m¡®àπ B–$lÄ®‰ZÿêÙïDÕ≤&-ú'Qi∑dÁê
L¿™O¥É≠ÄÅ|ìò‚Jíâx–áU(∫¶«®†TA•%ö∂Èè1B2–(L¡ƒŸVë™◊G™·"+8ÕC	RÄ#h!√QÄv–AHëMÌGÈGº"áÄjàW`ÅJ _≈$ÆÃÇ·7q±s‚"¬Y8ä`æqÑÄàt2~l¨Bç‰'@ÀŸ∞moZ1@≈z'ÜxˇÇ+à òu°1rb(ÃQ…ú
0Mã≈S=@¿JZÎQ[ΩKÅ&`ı”™xîMıgΩ„Åë FàAòygK≈lÈ†	ßàcD áÚ¨Ω∆ƒÊﬁz¥‘Äc≤"h‡
¨ÿdË`1·∏ëwi18Oä
ôÊàÍ“#+ó€Ds'ä¿	ê@$1æf#Z
5§
;)ËåJAQKI|Cª5D¥,#îiÀXÁ1¨úµçiã◊¢xãúléZÒGò·ƒH≤b‰rz
Ïn5N˝ƒ◊‘k∂I¸,dÖI¥Ì“¿ÌC“ XgÙµ>°	
ƒh∆kjÓlpaJ ¿ß˙¿∞®„≥!HEI§@®.LEê«açÄ∞∫Q	Ír√Œ∏ ©éáã¥Ì§ò%öÉπ€	bœ† !Yóçä>ùO Ä÷œÓ`&ênﬁ
,eSkd‰“%®Ohå≠ZO\vÿm†÷R  \ñæc∏üC¯≤|‰¥d-äÄ A∆ì@br8é EW89{sJF`saBËÉDº§õç%ö@RJ’√;)gy’ÃBEÀ…[Bêî¢¡_â)Uîû!HeÓñ
Â»Õ#´KÉwÅ®ôXv| ãÍc7/∆:syÉÄ·RëI∞tÁ/Ç#púyÕ$ò–Œ(„•4B„
`êT¿;£¢Úìj«8Vq¡èΩáΩ*f™ÒPËS†.XêCÇ¢vmM– Õ&îçé.mYàÕ 1'`®]}#B -h°âé“\ŸôIK‡DZÂ
ÄF=zYiœFH\∂f¯38«⁄@+s5À†*2·*6 dX‹ I
(√ò\ã$S,åX©Jy
